Add label to client

danielberbece opened the pull request:
Added a button to use for adding a label to a client

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -380,25 +380,34 @@ export class ClientPageStore extends ComponentStore<ClientPageState> {
+  readonly addClientLabel = this.effect<string>(
+    obs$ => obs$.pipe(
+      withLatestFrom(this.selectedClientId$),
+      map(([label, clientId]) => {
```
In your code, `map` returns an Observable that would send the HTTP POST request, but this inner observable is not subscribed to. Try `switchMap` instead of `map`. Further reading: https://medium.com/@luukgruijs/understanding-rxjs-map-mergemap-switchmap-and-concatmap-833fc1fb09ff

danielberbece replied:
Thanks, that was the problem!

danielberbece commented:
I updated the pull-request. I implemented an `Add label` button, which upon click shows a dialog where the user can enter the new label. When a label is added, the client page is updated and the new label appears. 
The tests for this feature are still a work in progress.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ng.html:
```diff
@@ -0,0 +1,24 @@
+      </mat-option>
+    </mat-autocomplete>
+
+    <mat-error *ngIf="newLabelInputControl.invalid">This label is already present!</mat-error>
```
Use `hasError('alreadyPresentLabel')` instead of `invalid` to only show this error message for the validator that should trigger it.

danielberbece replied:
Good idea!

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ng.html:
```diff
@@ -0,0 +1,24 @@
+</div>
+<div mat-dialog-actions class="actions">
+  <button mat-button color="primary" (click)="onCancelClick()">Cancel</button>
+  <button mat-button color="primary" (click)="onAddClick()">Add label</button>
```
Since the onclick-handler does nothing when the form is invalid, this button should be disabled when the form is disabled. Since adding a label can be regarded as an idempotent action (the collection of client labels has the same state, even if you add a label twice), I'd also be fine with not handling this case at all and allowing the user to try to add a label twice.

danielberbece replied:
I think that disabling the button would be a nice touch!

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ng.html:
```diff
@@ -0,0 +1,24 @@
+      [matAutocomplete]="auto">
+
+    <mat-autocomplete #auto="matAutocomplete">
+      <mat-option *ngFor="let option of filteredCurrentLabels | async" [value]="option">
```
Is this autocomplete only suggesting already pre-existing labels of the client?

danielberbece replied:
Exactly!

max-vogler replied:
We want to suggest all labels in the autocomplete, not only labels that are already assigned to the client.

danielberbece replied:
All labels from all clients? Is there an API for this?

max-vogler replied:
Yes, use `ApiListClientsLabelsHandler` via `GET /api/v2/clients/labels`.

To clarify: Clients and Labels have a many-to-many relationship. Any Client can have `[0,n]` labels assigned and any label be assigned to many clients. As an implementation detail, if a label is no longer assigned to any client, it is also deleted from the database, but that does not need to concern you here.

When adding a label to a Client, the user can type in an arbitrary string to be used as label. The autocomplete should suggest any pre-existing labels, not only labels that are already assigned to the current client.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ts:
```diff
@@ -0,0 +1,55 @@
+      const input = control.value.toLowerCase();
+      const matchingLabels = this.clientLabels
+        .map(clientLabel => clientLabel.name)
+        .filter(label => label.toLowerCase() === input);
```
Use `find()` over() `filter` to  reduce the average runtime.

danielberbece replied:
Good idea!

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/api_interfaces.ts:
```diff
@@ -36,6 +36,14 @@ export declare interface ApiClientLabel {
+/**
+ * AddClientLabel proto mapping.
+ */
+export declare interface ApiAddClientLabel {
```
Please rename to the correct `ApiAddClientsLabelsArgs`.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ts:
```diff
@@ -0,0 +1,97 @@
+})
+export class ClientAddLabelDialog {
+  constructor(
+    readonly dialogRef: MatDialogRef<ClientAddLabelDialog>,
+    @Inject(MAT_DIALOG_DATA) private clientLabels: ReadonlyArray<ClientLabel>,
```
Is it possible to change both to `private readonly`?

danielberbece replied:
Totally

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ts:
```diff
@@ -0,0 +1,97 @@
+    .pipe(
+      filter((input): input is string | String => input !== undefined),
+      map(input => input.trim()),
+      combineLatest(this.allClientsLabels$.pipe(
```
I don't think combineLatest works as you intended here. The [docs](https://www.learnrxjs.io/learn-rxjs/operators/combination/combinelatest) do not show it used in a `pipe`. Instead, I think you want to use `withLatestFrom`.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ts:
```diff
@@ -0,0 +1,97 @@
+    .pipe(
+      map(([input, allLabels]) => {
+        if (input === '') return false;
+        return allLabels.find(label => input === label) === undefined;
```
Use `allLabels.includes()`, which better expresses what you try to achieve.

max-vogler commented on grr/server/grr_response_server/gui/ui/components/client_add_label_dialog/client_add_label_dialog.ts:
```diff
@@ -0,0 +1,97 @@
+        if (input === '') return [];
+        return allLabels.filter(label => label.includes(input));
+      }),
+      map(filteredLabels => filteredLabels.filter((label) => this.clientHasLabel(label) === false))
```
Use `!` instead of `=== false`.

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/api_interfaces.ts:
```diff
@@ -36,6 +36,17 @@ export declare interface ApiClientLabel {
   readonly name?: string;
 }
 
+export declare interface ApiClientsLabels {
```
Please use the exact names from the Protobuf files (`ApiListClientsLabelsResult` see proto/grr_response_proto/api/client.proto). This allows us to switch from the manual definitions to auto-generated ones in the near future.

max-vogler replied:
Also, please add a one-line JSDoc. I know it doesn't really contain information, but it is a necessary evil to keep the linter happy until we have code generation in place.

danielberbece replied:
Got it

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/api_interfaces.ts:
```diff
@@ -36,6 +36,17 @@ export declare interface ApiClientLabel {
+ * AddClientsLabelsArgs proto mapping.
+ */
+export declare interface ApiAddClientsLabelsArgs {
+  readonly client_ids: string[];
```
The API expects lowerCamelCase names. What is `client_ids` in the Protobuf should be called `clientIds` in TypeScript.

max-vogler replied:
Please use optional ReadonlyArray for all repeated Protobuf fields, e.g. `readonly clientIds?: ReadonlyArray<string>`. Generally, we try to keep data structures mostly immutable to leverage advantages of function programming (e.g. no unexpected modifications through side-effects, easy to understand code).

danielberbece replied:
Is there an exception to this rule?

max-vogler replied:
Yeah, we are not 100 % strict about the immutability. Sometimes the type system does not permit, e.g. NgRx expects a mutable `string[]`.

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api/http_api_service.ts:
```diff
@@ -215,6 +215,17 @@ export class HttpApiService {
+
+  addClientLabel(clientId: string, label: string): Observable<ApiAddClientsLabelsArgs> {
+    const url = `${URL_PREFIX}/clients/labels/add`;
+    return this.http.post<ApiAddClientsLabelsArgs>(url, {client_ids: [clientId], labels: [label]});
```
The generic type parameter in angle brackets specifies the return type, not the argument type. This API handler does not return a real payload (it only returns `{status: 'ok'}`), so you can use `{}` to specify that the HTTP response contains an object with unknown properties.

danielberbece replied:
That was a novice mistake, I must say

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api_translation/client.ts:
```diff
@@ -14,6 +14,15 @@ function createKnowledgeBase(kb: ApiKnowledgeBase): KnowledgeBase {
   };
 }
 
+export function translateClientLabel(label: ApiClientLabel): ClientLabel {
```
Please add a one-liner JSDoc to all exported symbols (functions, constants, ...).

max-vogler commented on grr/server/grr_response_server/gui/ui/lib/api_translation/client.ts:
```diff
@@ -14,6 +14,15 @@ function createKnowledgeBase(kb: ApiKnowledgeBase): KnowledgeBase {
+  let owner = '';
+
+  if (!label.name) throw new Error('name attribute is missing.');
+  if (label.owner) owner = label.owner;
```
The purpose of functions in api_translation is to convert API data structures that are heavily influenced by the Protobuf language to pure, plain-old TypeScript objects. In Proto best practices, all fields should be `optional`, because `required` fields can NEVER be changed to `optional` in the future.

Assigning the empty string to `label.owner` is a hack to satisfy the type system, but makes using ClientLabel hard. With an empty string as replacement for `undefined`, code now needs to check for the special value of `''` to really be sure that the ClientLabel has an owner.

Please instead, throw an Error if label.owner is false-y.

danielberbece replied:
I tried throwing an error when the `owner` property is undefined, but the result of this was an empty autocomplete list. What do you think could be the issue and a possible solution?

max-vogler replied:
Is the error occurring in tests or when you live test it in the browser? You can walk me through in a VC

danielberbece commented:
I updated the pull request. The API doesn't send the owner when requesting all clients labels, so I addressed this issue without affecting the current implementation of the ClientLabel. The PR is ready for review now.

max-vogler commented on grr/server/grr_response_server/gui/ui/store/config_facade.ts:
```diff
@@ -55,6 +62,14 @@ export class ConfigStore extends ComponentStore<ConfigState> {
+      obs$ => obs$.pipe(
+          switchMapTo(this.httpApiService.fetchAllClientsLabels()),
+          map(apiClientsLabels =>
+              apiClientsLabels.map(getApiClientLabelName)),
```
Please add a short comment why you only get the label name (stating that the API includes `owner` inconsistently).

max-vogler commented on grr/server/grr_response_server/gui/ui/store/client_page_facade.ts:
```diff
@@ -54,9 +54,9 @@ interface ClientPageState {
-      private readonly httpApiService: HttpApiService,
-      private readonly configService: ConfigService,
-      private readonly configFacade: ConfigFacade,
+    private readonly httpApiService: HttpApiService,
```
Your PR has hundreds of lines reformatted with whitespace changes. This pollutes the `git blame` view and makes it hard to track actual changes. Also, our internal formatter will override it again.

Please use [clang format](https://github.com/xaverh/vscode-clang-format-provider) to format your code. Use `"clang-format.style": "{BasedOnStyle: google}",` in the settings. If this does not work for you, please try https://github.com/google/gts.
